httpx/cert_reload.go

```diff
@@ -0,0 +1,159 @@
+package httpx
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"path"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ory/x/logrusx"
+)
+
+type (
+	CertLocation struct {
+		CertPath string
+		KeyPath  string
+	}
+
+	CertLoadFunc func() ([]tls.Certificate, error)
+
+	CertProvider struct {
+		certs   []tls.Certificate
+		mu      sync.Mutex
+		watcher *fsnotify.Watcher
+
+		certLoad     CertLoadFunc
+		certLocation CertLocation
+		logger       *logrusx.Logger
+	}
+)
+
+//NewCertProvider creates a tls.Certificate provider that reloads certificates on fs change
+//certs must not be empty
+func NewCertProvider(certs []tls.Certificate, certLocation CertLocation, certLoad CertLoadFunc, l *logrusx.Logger) *CertProvider {
```

Comment:
Personally, I find this interface a little bit confusing. We give a filesystem location to the function to watch, but we never actually read from the file (instead deferring to the `certLoad` callback). Do we need this complexity? Same with the initial `certs` argument. I think the interface would be better if it just accepted the location and did all reading internally.